PoX Lockup and Delegation - Change requests

lgalabru opened the pull request:
This branch was built upon `feat/pox-stacks-lockup` and is addressing the change requests voiced  in the PR https://github.com/blockstack/stacks-blockchain/pull/1790.
Commits were pushed in a different branch for optimising review.

While working on this PR, I was confronted to the fact that one of our integration test `integration_test_get_info` was silently failing pretty early in the process:
```
#12 33.98 test tests::integrations::integration_test_get_info ... INFO [1598490250.798] [src/net/http.rs:2565] [ThreadId(28)] Failed to parse HTTP request: ClientError(Message("Failed to deserialize posted transaction: Failed to deserialize contract name: too short or too long: 4"))
1167
```
This PR is fixing this issue.

kantai commented on src/vm/ast/parser/mod.rs:
```diff
@@ -8,7 +8,7 @@ use vm::errors::{RuntimeErrorType, InterpreterResult as Result};
 use vm::types::{Value, PrincipalData, TraitIdentifier, QualifiedContractIdentifier};
 
-pub const CONTRACT_MIN_NAME_LENGTH : usize = 5;
+pub const CONTRACT_MIN_NAME_LENGTH : usize = 1;
```
🥇

kantai reviewed (approved):
LGTM!

jcnelson commented on src/vm/contexts.rs:
```diff
@@ -671,10 +664,20 @@ impl <'a,'b> Environment <'a,'b> {
+            return result
         }
+
+        let result = self.global_context.handle_tx_result(result);
```
I think this addresses your earlier comment about the special-cases code path not working within nested `contract-call?`'s?  If so, can you add a test case to the PoX tests that tests nested invocations of the PoX lock-up contract?

jcnelson replied:
I haven't seen a test for this yet.  Is there one?

lgalabru replied:
Test case added with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/4374a9cbd9fd4fd3dd992e25b1bb57369ffe0de4, and I actually reverted my change, it turns out your initial approach was correct as well.

jcnelson commented on src/vm/database/structures.rs:
```diff
@@ -93,3 +93,114 @@ clarity_serializable!(i128);
+            debug!("Consolidated after account-token-lock");
+        }
+
+        // todo(ludo): Also check that unlock_height > current_height
```
Please address this.

lgalabru replied:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/7cf92506281c3a4e4c77f54f741ecd7999d8d03f

jcnelson commented on src/vm/docs/mod.rs:
```diff
@@ -85,10 +85,10 @@ contract principal.",
 
 const TOTAL_LIQUID_USTX_KEYWORD: KeywordAPI = KeywordAPI {
-    name: "total-liquid-ustx",
+    name: "stx-liquid-supply",
```
I had `ustx` because the value is in _micro_ STX.

lgalabru replied:
Yeah I think I've also raised this in the past.
All the other clarity **stx** functions are dealing with micro STX (stx-get-balance, stx-transfer).
We should either use the **ustx** prefix in the other functions, or stick with **stx**, but a mix of the two is very confusing.

jcnelson replied:
Fair point :+1:

jcnelson commented on src/vm/functions/assets.rs:
```diff
@@ -103,19 +93,13 @@ pub fn stx_transfer_consolidated(env: &mut Environment, from: &PrincipalData, to
     // loading from's locked amount and height
     // TODO: this does not count the inner stacks block header load, but arguably,
     // this could be optimized away, so it shouldn't penalize the caller.
+    // todo(ludo): double check this
```
There should be additional memory accounting for loading the PoX lock state.

lgalabru replied:
ha, do you mind unpacking you thoughts? thanks!

jcnelson replied:
The lock state isn't 0 bytes.  The sender should be paying to load whatever lock state is necessary to carry out the token transfer.  Ideally, the sender does _not_ pay for loading any receiver lock state (since before, the sender only had to increase the receiver's unlocked token balance).  But, the sender _always_ pays for loading its own lock state, and if it needs to re-consolidate its now-unlocked tokens, it _may_ pay for storing its new lock state as well as its new balance.

jcnelson replied:
Now that we've reached consensus on using a single MARF read/write and a 48-byte `STXBalance` struct, this code should call `add_memory()` to accurately reflect the size of the `STXBalance` instances loaded.  Right now, it's undercounting the memory use.

lgalabru replied:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/0c76e29ed9431ef3950a5ccf558f0dc568a617fc

jcnelson commented on src/chainstate/stacks/db/blocks.rs:
```diff
@@ -3365,14 +3367,16 @@ impl StacksChainState {
+
         // 5: the paying account must have enough funds
-        if fee as u128 > payer.stx_balance {
+        if  !payer.stx_balance.can_transfer(fee as u128, 0) {
```
This shouldn't be 0.  It should be the current block height, since the payer may have tokens that are locked that would be unlocked in the act of spending a transaction fee.

lgalabru replied:
Good catch, thanks! I had a unresolved todo a few lines above for this issues.

lgalabru replied:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/0c6c1d9dd8e47e79643c1edea6dc95de404ed262

jcnelson commented on src/chainstate/stacks/db/blocks.rs:
```diff
@@ -3392,8 +3396,8 @@ impl StacksChainState {
                     };
-                if total_spent > origin.stx_balance {
-                    return Err(MemPoolRejection::NotEnoughFunds(total_spent, origin.stx_balance))
+                if !origin.stx_balance.can_transfer(total_spent, 0) {
```
Same problem here.

lgalabru replied:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/0c6c1d9dd8e47e79643c1edea6dc95de404ed262

jcnelson commented on src/chainstate/stacks/db/blocks.rs:
```diff
@@ -3392,8 +3396,8 @@ impl StacksChainState {
-                if total_spent > origin.stx_balance {
-                    return Err(MemPoolRejection::NotEnoughFunds(total_spent, origin.stx_balance))
+                if !origin.stx_balance.can_transfer(total_spent, 0) {
+                    return Err(MemPoolRejection::NotEnoughFunds(total_spent, origin.stx_balance.get_available_balance_at_block(0)))
```
And here.

lgalabru replied:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/0c6c1d9dd8e47e79643c1edea6dc95de404ed262

jcnelson commented on src/clarity.rs:
```diff
@@ -425,10 +425,11 @@ pub fn invoke_command(invoked_by: &str, args: &[String]) {
-                        let final_balance = cur_balance.checked_add(*amount as u128).expect("FATAL: account balance overflow");
-                        db.set_account_stx_balance(principal, final_balance as u128);
-                        println!("{} credited: {} uSTX", principal, final_balance);
+                        let mut balance = STXBalance::zero();
```
Should there be a `STXBalance::initial(x)`-like method that instantiates a balance with an initial value?  It seems weird that you're doing a separate `.credit()` operation, and having to pass in a magic constant `0` for the block height.

lgalabru replied:
Good idea

lgalabru replied:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/a15d408578a925f9c77029f0480668fa87161e5a

jcnelson commented on src/vm/clarity.rs:
```diff
@@ -1118,13 +1118,12 @@ mod tests {
 
         tx3.post_conditions.push(TransactionPostCondition::STX(PostConditionPrincipal::Origin, FungibleConditionCode::SentEq, 100));
-
+        let mut stx_balance = STXBalance::zero();
```
Same here.

lgalabru replied:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/a15d408578a925f9c77029f0480668fa87161e5a

jcnelson commented on src/vm/database/structures.rs:
```diff
@@ -93,3 +93,114 @@ clarity_serializable!(i128);
+        Ok(())
+    }
+
+    pub fn credit(&mut self, amount: u128, block_height: u64) -> Result<()> {
```
Why does `credit` take a block height?  Also, why does crediting tokens also unlock them?  That seems unfair to the token sender -- it forces the token sender to pay to unlock the recipient's tokens.

lgalabru replied:
Yes, because previously, unlocking tokens for the recipient would have a cost (you'd have to update the MARFwith the new `locked_amount `, `unlock_height`). 
Now that `(unlocked_amount, locked_amount, unlock_height)` are stored under the same MARF key, we can use the `unlock_amount` update as a vehicle for also updating the values `locked_amount `, `unlock_height`.

jcnelson replied:
>  unlocking tokens for the recipient would have a cost

No, it wouldn't.  The _sender_ would pay to unlock its tokens before sending them to the recipient.  In the previous version, the recipient's unlocked balance would increment, but the lock state for the recipient wouldn't change.

lgalabru replied:
Yes that's what I'm saying

> it forces the token sender to pay to unlock the recipient's tokens. I intentionally did not do this originally

With the previous approach, materializing the unlock in the MARF for the recipient would cost to the sender, and this would have been unfair for the sender. 
After addressing the change request https://github.com/blockstack/stacks-blockchain/pull/1790#discussion_r479542332, updating the lock for the recipient is no longer an extra cost for the sender.

jcnelson replied:
> With the previous approach, materializing the unlock in the MARF for the recipient would cost to the sender, and this would have been unfair for the sender.

The sender paid only to unlock its own tokens in the original PR.  That was intentional.  The recipient's lock state was not touched.

Here is the reproduced `stx_transfer_consolidated` method, to show you what I mean:

```
/// Do a "consolidated" STX transfer.
/// If the 'from' principal has locked STX, and they have unlocked, then process the STX unlock
/// and update its balance in addition to spending tokens out of it.
pub fn stx_transfer_consolidated(env: &mut Environment, from: &PrincipalData, to: &PrincipalData, amount: u128) -> Result<Value> {
    if amount <= 0 {
        return clarity_ecode!(StxErrorCodes::NON_POSITIVE_AMOUNT);
    }

    if from == to {
        return clarity_ecode!(StxErrorCodes::SENDER_IS_RECIPIENT);
    }

    if Some(from.clone()) != env.sender.as_ref().map(|pval| pval.clone().expect_principal()) {
        return clarity_ecode!(StxErrorCodes::SENDER_IS_NOT_TX_SENDER);
    }

    let (from_bal, unlock) = stx_balance_consolidated(&mut env.global_context.database, from);
    let to_bal = env.global_context.database.get_account_stx_balance(to);

    if from_bal < amount {
        return clarity_ecode!(StxErrorCodes::NOT_ENOUGH_BALANCE)
    }

    let final_from_bal = from_bal - amount;
    let final_to_bal = to_bal.checked_add(amount)
        .ok_or(RuntimeErrorType::ArithmeticOverflow)?;

    // loading from/to principals and balances
    env.add_memory(TypeSignature::PrincipalType.size() as u64)?;
    env.add_memory(TypeSignature::PrincipalType.size() as u64)?;
    env.add_memory(TypeSignature::UIntType.size() as u64)?;
    env.add_memory(TypeSignature::UIntType.size() as u64)?;

    // loading from's locked amount and height
    // TODO: this does not count the inner stacks block header load, but arguably,
    // this could be optimized away, so it shouldn't penalize the caller.
    env.add_memory(TypeSignature::UIntType.size() as u64)?;
    env.add_memory(TypeSignature::UIntType.size() as u64)?;

    // NOTE: this updates the balance with the unlocked tokens, if we did an unlock.
    env.global_context.database.set_account_stx_balance(from, final_from_bal);
    env.global_context.database.set_account_stx_balance(to,   final_to_bal);

    if unlock {
        // clear lock state
        env.global_context.database.set_account_stx_locked(from, 0);
        env.global_context.database.set_account_unlock_height(from, 0);
        debug!("Consolidated {} after stx-transfer", from);
    }

    env.global_context.log_stx_transfer(&from, amount)?;
    env.register_stx_transfer_event(from.clone(), to.clone(), amount)?;
    Ok(Value::okay_true())
}
```

Notice that only the _sender's_ lock state is loaded and stored.  The only data loaded and stored for the recipient (`to`) is the (unlocked) token balance.  That was intentional.

Per https://github.com/blockstack/stacks-blockchain/pull/1790#discussion_r479542332, it would be fine to consolidate both the STX locked and block height into a single MARF key/value pair, since they are always loaded/stored together anyway.  But, I do not think the three fields should be combined, since that would penalize the sender, who would then be forced to at least load the lock state for a recipient _even though it would not be used_.   I'll update the comment to reflect this as well.

Even if we did consolidate all three fields and (IMHO unnecessarily) penalize the sender for sending to a recipient, in no case should the sender be paying to unlock the recipient's tokens.  Storing data is even more costly than loading it.

lgalabru replied:
I should have phrase differently :). I noticed and **understood** that in your first implementation, you **did not** materialize the lock update, because it **would** have cost to the sender, and this would have been unfair.
Addressing your initial question: I took a different approach in this implementation, and did chose to update the lock because the balance's recipient (including unlocked_amount, locked_amount and unlock_height) need to be get and set, since we're updating the unlocked_amount.

jcnelson replied:
I'm glad we're on the same page :) 

> Addressing your initial question: I took a different approach in this implementation, and did chose to update the lock because the balance's recipient (including unlocked_amount, locked_amount and unlock_height) need to be get and set, since we're updating the unlocked_amount.

I consider this a design flaw, then.  The sender should not be paying to unlock the recipient's tokens.  Doing so would increase the runtime read cost of a `TokenTransfer` by 33%, and write cost by up to 33%

Before: load 4 u128's (sender balance, locked tokens, unlock block height; recipient balance), store either 2 or 4 u128s (sender and recipient balance, _possibly also_ sender locked tokens and sender unlock block height if the sender's tokens unlocked).

After:  load 6 u128's (sender balance, locked tokens, unlock block height; recipient balance, locked tokens, unlock block height); store either 2, 4, or 6 u128's (sender and recipient balance, _possibly also_ sender locked tokens and sender unlock block height if the sender's tokens unlocked, and _possibly also_ the recipient locked tokens and recipient unlock block height if the recipient's tokens unlocked).

It may not seem like much, but in aggregate it raises transaction fees for everyone.  All of these I/O operations come out of the block's total compute budget.

lgalabru replied:
I feel like there's a tradeoff to be made, and we should look at the big picture. We have 2 main use cases:

STX-TRANSFER (require a transaction fee)
- Before: read 4 MARF entries of 16 bytes, write 4 MARF entries of 16 bytes
- After: read 2 MARF entries of 48 bytes, write 2 MARF entries of 48 bytes

STX-GET-BALANCE (free of charge)
- Before: read 3 MARF entries of 16 bytes
- After: read 1 MARF entry of 48 bytes

We should consider the fact that the ratio of calls `stx-transfer`:`stx-get-balance` is far from 1:1 (1:10? 1:100?).

This feedback made sense when I started addressing it, because I was **under the impression** that MARF reading 3 entries of 16 bytes was more expensive than MARF reading 1 entry of 48 bytes.

kantai replied:
I agree with @lgalabru here -- the additional 32 bytes read and serialized out of the database are _very_ small compared to the cost of additional MARF reads or writes.

jcnelson replied:
I think something got lost here.  I'm talking about the costs that the VM tracks for calculating how full a block is.  The fundamental issue this comment thread is about is making sure that we keep the VM-accounted runtime costs as low as possible for the transaction sender.  Unless I'm missing something, this change is making it worse.

@kantai are you saying that 1 MARF read for 32 bytes is tabulated by the VM as a cheaper operation than 3 MARF reads for 16 bytes?

jcnelson replied:
Resolution: we will weight MARF read/write operations more heavily than byte read/write operations, since they are considerably more expensive.  This would mean that 1 MARF read for 48 bytes would be considerably cheaper to the VM than 3 MARF reads for 16 bytes.

jcnelson replied:
From the public blockchain meeting, we'll keep the current STX balance representation, and we'll be sure to weight MARF I/O operations much more heavily (i.e. consider them much more expensive than) raw I/O on the sidestore (https://github.com/blockstack/stacks-blockchain/issues/1877).

jcnelson commented on src/vm/contexts.rs:
```diff
@@ -452,8 +452,9 @@ impl <'a> OwnedEnvironment <'a> {
-                                let bal = env.global_context.database.get_account_stx_balance(recipient);
-                                env.global_context.database.set_account_stx_balance(recipient, bal + amount);
+                                let mut balance = env.global_context.database.get_account_stx_balance(recipient);
+                                balance.credit(amount, 0).unwrap();
```
Why is the block height 0?

lgalabru replied:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/189314ca993c9a3b881a773c45b974326f45e96a

jcnelson commented on src/vm/database/structures.rs:
```diff
@@ -93,3 +93,114 @@ clarity_serializable!(i128);
+        }
+    }
+
+    pub fn lock_tokens(&mut self, amount_to_lock: u128, unlock_height: u64, current_height: u64) -> Result<()> {
```
So, the reason the lock/unlock logic was all inlined with database loads/stores in the previous PR was because I didn't want there to be any unnecessary representations (or additional mutable representations) of an account's token lock state.  Before, there were only two representations:  the _read-write_ database, and a _read-only_ copy of the account as it was loaded.  This was intentional, because it made it _impossible_ to lock tokens _without also_ locking them in the database.

This method, and this representation of STX balance and lock state, makes me deeply uncomfortable.  It's possible to write buggy code _that will compile_ that calls this method and one of the unlock methods multiple times, and in doing so, create a representation of an account's balance within the `STXBalance` instance that is not consistent with the blockchain transactions' token movements.

I think any methods here that take `&mut self` should be removed, and the necessary logic they contain re-inlined with the database operations that load/store account balances and lock state.

lgalabru replied:
> buggy code that will compile that calls this method and one of the unlock methods multiple times, and in doing so, create a representation of an account's balance within the STXBalance instance that is not consistent with the blockchain transactions' token movements

Do you think you can write a snippet of code that would illustrates and give more context? That would be helpful. In my mind, if a developer wants an immutable instance of the balance, they will just declare the var as non mutable, and won't be able to call the mutating functions? 

I think it's in our interest to DRY as much as possible the critical sections of our codebase, and I would qualify the balance logic as critical. 
I felt uncomfortable when I saw that the **logic** around the balance was inlined, repeated, and coupled with the MARF (which increase the difficulty to test it). That was perfectly fine when the only thing we were doing was `checked_add` and `checked_sub`, but now that the logic around locks is introducing subtleties, it might be time to rely on some helpers (and I feel like it already paid off, since this refactoring helped catching issues that we missed during the review).

jcnelson replied:
The invariant that I want to be enforced here is that anytime you alter the lock state for an account, you _also_ do so in the database.  This invariant was enforced by the previous PR, but it is no longer enforced here.

To see why this invariant is important, consider this hypothetical code path with the new methods:

```
let mut account = StacksChainState::get_account(clarity_tx, alice)?;
account.stx_balance.lock_tokens(alice_lockup, alice_unlock_height, cur_block_height)?;

/* At this point, account.stx_balance and the Clarity DB are _inconsistent _. */
/* To be safe, you should _immediately_ save the new lock state. */
/* If you fail to do so, then suppose now that this code subsequently runs: */
let mut spending_account = StacksChainState::get_account(clarity_tx, alice)?;
spending_account.debit(amount, cur_block_height)?;  // MAY SUCCEED, even if _all_ of Alice's tokens were supposed to be locked!
```

With my original PR, the attempt to debit `alice`'s account is guaranteed to fail, because the second call to `get_account` would always read the latest locked state.  This property was enforced by ensuring that the only way to lock tokens was by also altering the database.

With this new PR, it is possible to accidentally write code that succeeds in debiting `alice`'s account, because you forgot to also save the lock state.  While it is possible to avoid the above bug, it's _harder_ to do so, because now each time you call `lock_tokens` (or do some other mutation on `STXBalance`), you have to manually ensure that each such instance is coupled with a corresponding set of calls to mutate the state in the Clarity database.  The point I'm trying to emphasize here is that **this is more dangerous than what we had been doing before**, because it makes correctness dependent on us _humans_ always manually verifying that all mutations are materialized, even months and years later when we're maintaining this code (whereas before, we didn't have to think about this at all).

I'm not opposed to helper methods and DRY'ing code -- in fact, I encourage both.  But not at the expense of correctness (goes without saying), and not at the expense of making the codebase more dangerous to work with.

jcnelson replied:
Per the public blockchain call, we can leave this as is for now.  But, I intend to address this later, post Krypton.  https://github.com/blockstack/stacks-blockchain/issues/1878

lgalabru commented on src/vm/database/structures.rs:
```diff
@@ -93,3 +93,114 @@ clarity_serializable!(i128);
+clarity_serializable!(STXBalance);
+
+#[derive(Debug)]
+pub enum STXBalanceError {
```
Yeah I thought about this, I ended up optimizing for modularity by adding a custom error type, so that the caller can panic (or not) with a specific and contextual message.

jcnelson replied:
All instances of these errors are either converted into `.unwrap()` or `.expect()` by their callers, or converted to an `ArthithmeticError` without consideration of the error type.  So, I don't buy the argument for introducing this type if it's not going to be used in a meaningful way.  But, if you insist on keeping it, then it should not be `pub` until there is at least one other module using it.

lgalabru replied:
Making this enum private is not an option, since
```
pub fn debit(&mut self, amount: u128, block_height: u64) -> Result<()>
pub fn credit(&mut self, amount: u128, block_height: u64) -> Result<()>
pub fn transfer_to(&mut self, recipient: &mut STXBalance, amount: u128, block_height: u64) -> Result<()>
```
would end up leaking a private type, leading to a compiler error.
We could return an optional instead which looks like a downgrade.

jcnelson commented on src/vm/clarity.rs:
```diff
@@ -1118,13 +1118,12 @@ mod tests {
         tx3.post_conditions.push(TransactionPostCondition::STX(PostConditionPrincipal::Origin, FungibleConditionCode::SentEq, 100));
-
+        let mut stx_balance = STXBalance::zero();
+        stx_balance.credit(5000, 0).unwrap();
```
Why is the block height 0 here?

jcnelson commented on src/vm/contexts.rs:
```diff
@@ -452,8 +452,10 @@ impl <'a> OwnedEnvironment <'a> {
-                                env.global_context.database.set_account_stx_balance(recipient, bal + amount);
+                                let mut balance = env.global_context.database.get_account_stx_balance(recipient);
+                                let block_height = env.global_context.database.get_current_burnchain_block_height();
+                                balance.credit(amount, block_height as u64).unwrap();
```
Should this be `expect()`?

lgalabru replied:
This function is only available in `tests`, but sure. Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/0c76e29ed9431ef3950a5ccf558f0dc568a617fc

kantai commented:
@lgalabru -- should this PR also change its base branch to `next`?

lgalabru commented:
@kantai yes it should, I initially opened PR against original PR, to make the change request reviews easier.
I've just fixed the conflicts, and also added the commits dealing with the integration (https://github.com/blockstack/stacks-blockchain/pull/1874 -> https://github.com/blockstack/stacks-blockchain/pull/1874/commits/ec507d7c5997ed41a3240e1606b82a43748ebc84 / https://github.com/blockstack/stacks-blockchain/pull/1874/commits/9f5f398ce8517aaf395201b8c720f8b01b0975c3 / https://github.com/blockstack/stacks-blockchain/pull/1874/commits/d35ef465e9fa2ce327a181117f8ca7933b9df075)

jcnelson reviewed (approved):
Looks good to me.  Thanks @lgalabru for taking this PR over!

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
